atlasdb-processors/src/main/java/com/palantir/processors/TypeToExtend.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.processors;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.PackageElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+
+import com.google.common.collect.Sets;
+
+final class TypeToExtend {
+    private TypeElement typeToExtend;
+    private PackageElement typePackage;
+    private Set<ExecutableElement> methods;
+    private Set<ExecutableElement> constructors;
+
+    TypeToExtend(PackageElement typePackage,
+            TypeElement typeToExtend,
+            TypeElement... supertypes) {
+
+        this.typeToExtend = typeToExtend;
+        this.typePackage = typePackage;
+
+        List<ExecutableElement> allMethods = extractMethods(typeToExtend);
+        for (TypeElement supertype : supertypes) {
+            allMethods.addAll(extractMethods(supertype));
+        }
+
+        Map<String, ExecutableElement> unifiedMethods = allMethods
+                .stream()
+                .collect(Collectors.toMap(ExecutableElement::toString, Function.identity(), (name1, name2) -> name1));
```

Comment:
The `(name1, name2) -> name1` function is the conflict resolution for this hash map. It says that, in case of methods with the same signature, just choose the 'first' one, since they're both the same. Methods with the same signature will occur when a subclass overrides a method.

This guarantees that we have a map of 'uniqueMethodName -> method', and we store all unique methods on the variable `methods`. I've renamed some methods around this area and added some comments. Let me know if it's clearer.